### PR TITLE
Week 2: Add comments to the portfolio

### PR DESCRIPTION
**Pull request by SoyOscarRH**
- Added a new component in the frontend to handle sending and seeing the comments
- Use datastore to store the comments and create 2 servlets to update the information

This final result is this:

Or by yourself at:
http://orosashernandez-sps-spring20.appspot.com/

### CODE REVIEW

File: Portfolio/src/main/java/com/google/sps/servlets/CreateCommentsServlet.java [hunk 1]
```diff
@@ -15,18 +15,31 @@
...
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String body = request.getReader().lines().collect(Collectors.joining());
+    Map data = new Gson().fromJson(body, Map.class);
+    String username = data.get("username").toString();
+    String message = data.get("message").toString();
+
+    Comment comment = new Comment(username, message);
+    comment.toDataStore();
```
Comment (ctrokid) [inline Portfolio/src/main/java/com/google/sps/servlets/CreateCommentsServlet.java]: Interesting, I like this idea of abstracting away database details from us here and leaving them up to the Comment class. However, my concerns come when the Database interaction could be expensive in nature, in which we'd actually want the database to be abstracted away from the model in this case.

Would it be a better design for `toDataStore` to return an `Entity` in which the caller could write to Datastore? Combining this with my coherence suggestion, you could make the `datastore` variable a private member of `CommentsServlet` and re-use it between doGet and doPost respectively.

Hopefully this makes sense!

File: Portfolio/src/main/java/com/google/sps/servlets/CreateCommentsServlet.java [hunk 2]
```diff
@@ -15,18 +15,31 @@
...
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
-@WebServlet("/data")
-public class DataServlet extends HttpServlet {
+import java.util.stream.Collectors;
+import com.google.gson.Gson;
+
+@WebServlet("/create_comment")
```
Comment (ctrokid) [inline Portfolio/src/main/java/com/google/sps/servlets/CreateCommentsServlet.java]: `@WebServlet("/comment")`

File: Portfolio/src/main/java/com/google/sps/servlets/CreateCommentsServlet.java [hunk 3]
```diff
@@ -15,18 +15,31 @@
...
+public class CreateCommentsServlet extends HttpServlet {
```
Comment (ctrokid) [inline Portfolio/src/main/java/com/google/sps/servlets/CreateCommentsServlet.java]: `CommentServlet` or `CommentsServlet`

File: Webpage/Code/Comments/index.tsx [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    const username = elements["username"].value as string
+    const message = elements["message"].value as string
+    elements["username"].value = ""
+    elements["message"].value = ""
+
+    const body = JSON.stringify({username,  message})
+
+    const response = await fetch('/create_comment', { method: 'POST', headers, body})
+    const serverSaid = (await response.json())
+    console.log(serverSaid)
```
Comment (ctrokid) [inline Webpage/Code/Comments/index.tsx]: Do you want to keep this logging in the code after submitting?

File: Portfolio/src/main/java/com/google/sps/servlets/GetCommentsServlet.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+
+@WebServlet("/get_comments")
+public class GetCommentsServlet extends HttpServlet {
```
Comment (ctrokid) [inline Portfolio/src/main/java/com/google/sps/servlets/GetCommentsServlet.java]: I like that you want to make these very coherent and pull them into two separate classes, however I think there's a minimal amount of logic and it actually makes sense to put the GetComments and CreateComments actions in the same file (see my other comment for how we could get benefits here).

I think we should move the `getComments` and `doPost` methods into the common `CommentServlet`. doPost here should be renamed to `doGet` as GET requests return data from the server while POST/PUT requests entail the user will be writing data to be persisted on the server.

File: Portfolio/src/main/java/com/google/sps/servlets/CreateCommentsServlet.java (hunk 1 above)
Comment (SoyOscarRH) [reply Portfolio/src/main/java/com/google/sps/servlets/CreateCommentsServlet.java]: I agree with your idea to abstract away database details, good design idea

File: Webpage/Code/Comments/index.tsx (hunk 4 above)
Comment (SoyOscarRH) [reply Webpage/Code/Comments/index.tsx]: Hahaha, no, totally forgot to delete that at the end 😝

File: Portfolio/src/main/java/com/google/sps/servlets/GetCommentsServlet.java (hunk 5 above)
Comment (SoyOscarRH) [reply Portfolio/src/main/java/com/google/sps/servlets/GetCommentsServlet.java]: I have not many experience with backend, so this comments are great feedback for me.
I knew that GET, POST, PUT, entail different actions but forgot to apply it here.

Also again, this change made the code a lot more compact, readable and concise, good design idea.

**Pull request merged**